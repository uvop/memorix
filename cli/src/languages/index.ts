import { assertUnreachable } from "src/core/utilities";
import { codegenTs } from "./ts/ts";
import { codegenPython } from "./python/python";

export enum Languages {
  typescript = "typescript",
  python = "python",
}

type CodegenByLanguageFn = (schema: string, language: Languages) => string;

export const codegenByLanguage: CodegenByLanguageFn = (scope, language) => {
  switch (language) {
    case Languages.typescript:
      return codegenTs(scope);
    case Languages.python:
      return codegenPython(scope);
    default:
      assertUnreachable(language);
      return "";
  }
};
