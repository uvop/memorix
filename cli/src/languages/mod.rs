mod python;
mod rust;
mod typescript;

use crate::{
    flat_schema::FlatValidatedSchema,
    parser::{Export, Language},
    validate::ValidatedSchema,
};
use std::collections::HashMap;

pub fn codegen_per_language(
    export: &Export,
    validated_schema: &ValidatedSchema,
    flat_validated_schema: &FlatValidatedSchema,
) -> HashMap<String, String> {
    export
        .files
        .as_ref()
        .unwrap_or(&vec![])
        .iter()
        .map(|file| {
            let path = file.path.to_string();
            let content = match file.language {
                Language::TypeScript => typescript::codegen(validated_schema),
                Language::Python => python::codegen(flat_validated_schema),
                Language::Rust => rust::codegen(flat_validated_schema),
            };
            (path, content)
        })
        .collect()
}
