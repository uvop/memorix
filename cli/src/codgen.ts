import path from "path";
import fs from "fs";
import { codegenByLanguage } from "./languages";
import { Block, BlockTypes, getBlocks } from "./core/block";

const getSchemaBlocksNoConfig = async (
  schemaFilePath: string,
  dirname?: string
): Promise<Block[]> => {
  const schemaPath =
    dirname !== undefined
      ? path.resolve(dirname, schemaFilePath)
      : path.resolve(schemaFilePath);
  const schemaFolder = path.dirname(schemaPath);
  const schema = await (await fs.promises.readFile(schemaPath)).toString();
  const blocks = getBlocks(schema);
  const blockConfig = blocks.find((x) => x.type === BlockTypes.config);
  if (
    !blockConfig ||
    blockConfig.type !== BlockTypes.config ||
    !blockConfig.extends
  ) {
    return blocks;
  }

  const blocksToAdd = await Promise.all(
    (Array.isArray(blockConfig.extends)
      ? blockConfig.extends
      : [blockConfig.extends]
    ).map(async (schemaExtendFilePath) =>
      getSchemaBlocksNoConfig(schemaFolder, schemaExtendFilePath)
    )
  );

  return [
    ...blocksToAdd.flat(),
    ...blocks.filter((x) => x.type !== BlockTypes.config),
  ];
};

export const codegen = async ({
  schemaFilePath,
}: {
  schemaFilePath: string;
}) => {
  const schemaPath = path.resolve(schemaFilePath);
  const schemaDirname = path.dirname(schemaPath);
  const schema = await (await fs.promises.readFile(schemaPath)).toString();
  const blockConfig = getBlocks(schema).find(
    (x) => x.type === BlockTypes.config
  );
  if (!blockConfig || blockConfig.type !== BlockTypes.config) {
    throw new Error(
      `Must set a Config block to top level schema "${schemaPath}"`
    );
  }
  const otherBlocks = await getSchemaBlocksNoConfig(schemaFilePath);
  const blocks = [blockConfig, ...otherBlocks];

  await Promise.all(
    Array.isArray(blockConfig.output)
      ? blockConfig.output
      : [blockConfig.output].map(async ({ language, file }) => {
          const filePath = path.resolve(schemaDirname, file);
          const code = codegenByLanguage(blocks, language);
          await fs.promises.writeFile(filePath, code);
        })
  );
};
