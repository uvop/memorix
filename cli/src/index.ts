#!/usr/bin/env node
import yargs from "yargs/yargs";
import { hideBin } from "yargs/helpers";
import { Languages } from "./languages";
import { codegen } from "./codgen";

yargs(hideBin(process.argv))
  .command(
    "codegen <schemaFilePath> <...files>",
    "Codegen memorix schema to code",
    (b) => {
      return b
        .positional("schemaFilePath", {
          describe: "Memorix schema file",
        })
        .positional("files", {
          describe: "Pairs of language to codegen and destination file",
        });
    },
    (argv) => {
      const { schemaFilePath, files } = argv as {
        schemaFilePath: string;
        files: string[];
      };
      const parsedFiles: Parameters<typeof codegen>[0]["files"] = [];
      for (let index = 0; index < files.length; index += 2) {
        const languageStr = files[index];
        const dist = files[index + 1];
        const language = Languages[languageStr];
        if (language === undefined) {
          throw new Error(
            `Unknown language, got "${languageStr}", expected one of "${Object.keys(
              Languages
            ).join(", ")}"`
          );
        }
        if (dist === undefined) {
          throw new Error(`Must pass path for "${language}" codegen`);
        }
        parsedFiles.push({
          dist,
          language,
        });
      }
      if (parsedFiles.length === 0) {
        throw new Error(`Didn't get any languages to codegen`);
      }
      codegen({
        schemaFilePath,
        files: parsedFiles,
      });
    }
  )
  .option("verbose", {
    alias: "v",
    type: "boolean",
    description: "Run with verbose logging",
  })
  .demandCommand(1)
  .parse();
