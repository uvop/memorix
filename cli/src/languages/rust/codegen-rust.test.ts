import { Languages } from "src/languages";
import { codegen } from "src/codgen";
import fs from "fs";

jest.mock("fs", () => {
  return {
    promises: {
      writeFile: jest.fn(),
      readFile: jest.fn(),
    },
  };
});

const codegenRust: (schema: string, moreConfig?: string) => Promise<string> = (
  schema,
  moreConfig
) =>
  new Promise((res, rej) => {
    const schemaWithOutput = `
  Config {
    output: {
      language: "${Languages.rust}"
      file: "generated.ts"
    }
    ${moreConfig}
  }

  ${schema}
  `;

    (fs.promises.readFile as jest.Mock).mockImplementation(async () => ({
      async toString() {
        return schemaWithOutput;
      },
    }));
    (fs.promises.writeFile as jest.Mock).mockImplementation(
      async (path, content) => {
        res(content);
      }
    );

    codegen({ schemaFilePath: "schema.memorix" }).catch(rej);
  });

describe("python codegen", () => {
  describe("model", () => {
    it("can generate from model", async () => {
      expect(
        await codegenRust(
          `
            Model User {
                id: int
                name: string
                age: float?
            }
          `
        )
      ).toMatchSnapshot();
    });
    it("can generate from 2 models", async () => {
      expect(
        await codegenRust(
          `
            Model User1 {
                id: int
            }
            Model User2 {
                name: string
                age: int?
            }
          `.trim()
        )
      ).toMatchSnapshot();
    });
    it("can generate from model within model", async () => {
      expect(
        await codegenRust(
          `
            Model User {
                id: int
                papa: {
                    name: string?
                }
            }
          `.trim()
        )
      ).toMatchSnapshot();
    });
    it("can generate from model within model within a model", async () => {
      expect(
        await codegenRust(
          `
            Model User {
                id: int
                papa: {
                    name: string?
                    mama: {
                        sick: boolean
                    }?
                    age: int
                }
            }
          `.trim()
        )
      ).toMatchSnapshot();
    });
    it("can generate from model with array", async () => {
      expect(
        await codegenRust(
          `
            Model User {
                id: int
                names: [string]
                children: [{
                  id: int
                  name: string?
                }?]?
            }
          `
        )
      ).toMatchSnapshot();
    });
    it("throws if model appears twice", (done) => {
      codegenRust(
        `
          Model User {
              id: int
          }
          Model User {
              id: int
          }
        `
      )
        .then(() => {
          done("Was supposed to throw");
        })
        .catch(() => {
          done();
        });
    });
  });
  describe("cache", () => {
    it("can generate with inline types", async () => {
      expect(
        await codegenRust(
          `
            Cache {
              user {
                key: int
                payload: string
              }
            }
          `
        )
      ).toMatchSnapshot();
    });
    it("can generate with inline types and convert to CamelCase", async () => {
      expect(
        await codegenRust(
          `
            Cache {
              user_by_number {
                key: int
                payload: {
                  name: string
                  age: int
                }
              }
            }
          `
        )
      ).toMatchSnapshot();
    });
    it("can have options", async () => {
      expect(
        await codegenRust(
          `
            Cache {
              user {
                key: int
                payload: string
                options: {
                  expire: {
                    value: 5
                    extendOnGet: true
                  }
                }
              }
            }
          `
        )
      ).toMatchSnapshot();
    });
    it("can generate with inline object type", async () => {
      expect(
        await codegenRust(
          `
            Cache {
              user {
                key: int
                payload: {
                  name: string
                  age: int?
                }?
              }
            }
          `
        )
      ).toMatchSnapshot();
    });
    it("can generate with no key", async () => {
      expect(
        await codegenRust(
          `
            Cache {
              user {
                payload: {
                  name: string
                  age: int?
                }?
              }
            }
          `
        )
      ).toMatchSnapshot();
    });
    it("can generate also with model", async () => {
      expect(
        await codegenRust(
          `
          Cache {
            adminId {
                payload: string?
            }
            user {
                key: string
                payload: User
            }
          }
        
          Model User {
            name: string
            age: int?
          }
          `
        )
      ).toMatchSnapshot();
    });
  });
  describe("pubsub", () => {
    it("can generate with inline types", async () => {
      expect(
        await codegenRust(
          `
            PubSub {
              message {
                key: int
                payload: string
              }
            }
          `
        )
      ).toMatchSnapshot();
    });
  });
  describe("task", () => {
    it("can generate with inline types", async () => {
      expect(
        await codegenRust(
          `
            Task {
              doIt {
                key: int
                payload: string
                returns: boolean
              }
            }
          `
        )
      ).toMatchSnapshot();
    });
    it("can have options", async () => {
      expect(
        await codegenRust(
          `
            Task {
              doIt {
                key: int
                payload: string
                returns: boolean
                options: {
                  takeNewest: true
                }
              }
            }
          `
        )
      ).toMatchSnapshot();
    });
  });
  describe("enum", () => {
    it("can generate", async () => {
      expect(
        await codegenRust(
          `
            Enum Animals {
              dog
              cat
              person
            }
          `
        )
      ).toMatchSnapshot();
    });
  });
  describe("config", () => {
    it("can generate", async () => {
      expect(
        await codegenRust(
          `
            DefaultOptions {
              cache: {
                expire: {
                  value: 5
                  extendOnGet: true
                }
              }
              task: {
                takeNewest: true
              }
            }
          `
        )
      ).toMatchSnapshot();
    });
    it("can be null", async () => {
      expect(
        await codegenRust(
          `
            DefaultOptions {
              cache: {
                expire: null
              }
            }
          `
        )
      ).toMatchSnapshot();
    });
  });
  describe("namespace", () => {
    it("can generate", async () => {
      expect(
        await codegenRust(
          `
            Namespace user {
              DefaultOptions {
                cache: {
                  expire: {
                    value: 5
                  }
                }
              }
              Cache {
                bio {
                  payload: {
                    description: string
                  }
                }
              }
            }
            Namespace spaceship {
              DefaultOptions {
                cache: {
                  expire: {
                    value: 5
                  }
                }
              }
              Cache {
                bio {
                  payload: string
                }
              }
            }
          `
        )
      ).toMatchSnapshot();
    });
  });
});
