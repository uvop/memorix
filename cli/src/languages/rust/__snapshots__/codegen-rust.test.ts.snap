// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rust codegen cache can generate also with model 1`] = `
"#![allow(dead_code)]
extern crate memorix_client_redis;

#[derive(Clone, memorix_client_redis::Serialize, memorix_client_redis::Deserialize, PartialEq, std::fmt::Debug)]
pub struct User {
    pub name: String,
    #[serde(skip_serializing_if = \\"Option::is_none\\")]
    pub age: Option<i32>,
}


#[derive(Clone)]
#[allow(non_snake_case)]
pub struct MemorixCache {
    pub adminId: memorix_client_redis::MemorixCacheItemNoKey<Option<String>>,
    pub user: memorix_client_redis::MemorixCacheItem<String, User>,
}

impl MemorixCache {
    fn new(memorix_base: memorix_client_redis::MemorixBase) -> Self {
        Self {
            adminId: memorix_client_redis::MemorixCacheItemNoKey::new(
                memorix_base.clone(),
                \\"adminId\\".to_string(),
                None
            ),
            user: memorix_client_redis::MemorixCacheItem::new(
                memorix_base.clone(),
                \\"user\\".to_string(),
                None
            ),
        }
    }
}


#[derive(Clone)]
#[allow(non_snake_case)]
pub struct Memorix {

    pub cache: MemorixCache,
}

const MEMORIX_NAMESPACE_NAME_TREE: &[&str] = &[];

impl Memorix {
    pub async fn new(redis_url: &str) -> Result<Memorix, Box<dyn std::error::Error>> {
        let memorix_base = memorix_client_redis::MemorixBase::new(
            redis_url,
            MEMORIX_NAMESPACE_NAME_TREE,
            None
        ).await?;
        Ok(Self {

            cache: MemorixCache::new(memorix_base.clone()),
        })
    }
}"
`;

exports[`rust codegen cache can generate with inline object type 1`] = `
"#![allow(dead_code)]
extern crate memorix_client_redis;

#[derive(Clone, memorix_client_redis::Serialize, memorix_client_redis::Deserialize, PartialEq, std::fmt::Debug)]
pub struct CacheUserPayload {
    pub name: String,
    #[serde(skip_serializing_if = \\"Option::is_none\\")]
    pub age: Option<i32>,
}


#[derive(Clone)]
#[allow(non_snake_case)]
pub struct MemorixCache {
    pub user: memorix_client_redis::MemorixCacheItem<i32, Option<CacheUserPayload>>,
}

impl MemorixCache {
    fn new(memorix_base: memorix_client_redis::MemorixBase) -> Self {
        Self {
            user: memorix_client_redis::MemorixCacheItem::new(
                memorix_base.clone(),
                \\"user\\".to_string(),
                None
            ),
        }
    }
}


#[derive(Clone)]
#[allow(non_snake_case)]
pub struct Memorix {

    pub cache: MemorixCache,
}

const MEMORIX_NAMESPACE_NAME_TREE: &[&str] = &[];

impl Memorix {
    pub async fn new(redis_url: &str) -> Result<Memorix, Box<dyn std::error::Error>> {
        let memorix_base = memorix_client_redis::MemorixBase::new(
            redis_url,
            MEMORIX_NAMESPACE_NAME_TREE,
            None
        ).await?;
        Ok(Self {

            cache: MemorixCache::new(memorix_base.clone()),
        })
    }
}"
`;

exports[`rust codegen cache can generate with inline types 1`] = `
"#![allow(dead_code)]
extern crate memorix_client_redis;

#[derive(Clone)]
#[allow(non_snake_case)]
pub struct MemorixCache {
    pub user: memorix_client_redis::MemorixCacheItem<i32, String>,
}

impl MemorixCache {
    fn new(memorix_base: memorix_client_redis::MemorixBase) -> Self {
        Self {
            user: memorix_client_redis::MemorixCacheItem::new(
                memorix_base.clone(),
                \\"user\\".to_string(),
                None
            ),
        }
    }
}


#[derive(Clone)]
#[allow(non_snake_case)]
pub struct Memorix {

    pub cache: MemorixCache,
}

const MEMORIX_NAMESPACE_NAME_TREE: &[&str] = &[];

impl Memorix {
    pub async fn new(redis_url: &str) -> Result<Memorix, Box<dyn std::error::Error>> {
        let memorix_base = memorix_client_redis::MemorixBase::new(
            redis_url,
            MEMORIX_NAMESPACE_NAME_TREE,
            None
        ).await?;
        Ok(Self {

            cache: MemorixCache::new(memorix_base.clone()),
        })
    }
}"
`;

exports[`rust codegen cache can generate with inline types and convert to CamelCase 1`] = `
"#![allow(dead_code)]
extern crate memorix_client_redis;

#[derive(Clone, memorix_client_redis::Serialize, memorix_client_redis::Deserialize, PartialEq, std::fmt::Debug)]
pub struct CacheUserByNumberPayload {
    pub name: String,
    pub age: i32,
}


#[derive(Clone)]
#[allow(non_snake_case)]
pub struct MemorixCache {
    pub user_by_number: memorix_client_redis::MemorixCacheItem<i32, CacheUserByNumberPayload>,
}

impl MemorixCache {
    fn new(memorix_base: memorix_client_redis::MemorixBase) -> Self {
        Self {
            user_by_number: memorix_client_redis::MemorixCacheItem::new(
                memorix_base.clone(),
                \\"user_by_number\\".to_string(),
                None
            ),
        }
    }
}


#[derive(Clone)]
#[allow(non_snake_case)]
pub struct Memorix {

    pub cache: MemorixCache,
}

const MEMORIX_NAMESPACE_NAME_TREE: &[&str] = &[];

impl Memorix {
    pub async fn new(redis_url: &str) -> Result<Memorix, Box<dyn std::error::Error>> {
        let memorix_base = memorix_client_redis::MemorixBase::new(
            redis_url,
            MEMORIX_NAMESPACE_NAME_TREE,
            None
        ).await?;
        Ok(Self {

            cache: MemorixCache::new(memorix_base.clone()),
        })
    }
}"
`;

exports[`rust codegen cache can generate with no key 1`] = `
"#![allow(dead_code)]
extern crate memorix_client_redis;

#[derive(Clone, memorix_client_redis::Serialize, memorix_client_redis::Deserialize, PartialEq, std::fmt::Debug)]
pub struct CacheUserPayload {
    pub name: String,
    #[serde(skip_serializing_if = \\"Option::is_none\\")]
    pub age: Option<i32>,
}


#[derive(Clone)]
#[allow(non_snake_case)]
pub struct MemorixCache {
    pub user: memorix_client_redis::MemorixCacheItemNoKey<Option<CacheUserPayload>>,
}

impl MemorixCache {
    fn new(memorix_base: memorix_client_redis::MemorixBase) -> Self {
        Self {
            user: memorix_client_redis::MemorixCacheItemNoKey::new(
                memorix_base.clone(),
                \\"user\\".to_string(),
                None
            ),
        }
    }
}


#[derive(Clone)]
#[allow(non_snake_case)]
pub struct Memorix {

    pub cache: MemorixCache,
}

const MEMORIX_NAMESPACE_NAME_TREE: &[&str] = &[];

impl Memorix {
    pub async fn new(redis_url: &str) -> Result<Memorix, Box<dyn std::error::Error>> {
        let memorix_base = memorix_client_redis::MemorixBase::new(
            redis_url,
            MEMORIX_NAMESPACE_NAME_TREE,
            None
        ).await?;
        Ok(Self {

            cache: MemorixCache::new(memorix_base.clone()),
        })
    }
}"
`;

exports[`rust codegen cache can have options 1`] = `
"#![allow(dead_code)]
extern crate memorix_client_redis;

#[derive(Clone)]
#[allow(non_snake_case)]
pub struct MemorixCache {
    pub user: memorix_client_redis::MemorixCacheItem<i32, String>,
}

impl MemorixCache {
    fn new(memorix_base: memorix_client_redis::MemorixBase) -> Self {
        Self {
            user: memorix_client_redis::MemorixCacheItem::new(
                memorix_base.clone(),
                \\"user\\".to_string(),
                Some(memorix_client_redis::MemorixOptionsCache {
                    expire: Some(memorix_client_redis::MemorixOptionsCacheExpire {
                        value: 5,
                        is_in_ms: None,
                        extend_on_get: Some(true),
                    }),
                })
            ),
        }
    }
}


#[derive(Clone)]
#[allow(non_snake_case)]
pub struct Memorix {

    pub cache: MemorixCache,
}

const MEMORIX_NAMESPACE_NAME_TREE: &[&str] = &[];

impl Memorix {
    pub async fn new(redis_url: &str) -> Result<Memorix, Box<dyn std::error::Error>> {
        let memorix_base = memorix_client_redis::MemorixBase::new(
            redis_url,
            MEMORIX_NAMESPACE_NAME_TREE,
            None
        ).await?;
        Ok(Self {

            cache: MemorixCache::new(memorix_base.clone()),
        })
    }
}"
`;

exports[`rust codegen config can be null 1`] = `
"#![allow(dead_code)]
extern crate memorix_client_redis;

#[derive(Clone)]
#[allow(non_snake_case)]
pub struct Memorix {


}

const MEMORIX_NAMESPACE_NAME_TREE: &[&str] = &[];

impl Memorix {
    pub async fn new(redis_url: &str) -> Result<Memorix, Box<dyn std::error::Error>> {
        let memorix_base = memorix_client_redis::MemorixBase::new(
            redis_url,
            MEMORIX_NAMESPACE_NAME_TREE,
            Some(memorix_client_redis::MemorixOptions {
                cache: Some(memorix_client_redis::MemorixOptionsCache {
                    expire: None,
                }),
                task: None,
            })
        ).await?;
        Ok(Self {


        })
    }
}"
`;

exports[`rust codegen config can generate 1`] = `
"#![allow(dead_code)]
extern crate memorix_client_redis;

#[derive(Clone)]
#[allow(non_snake_case)]
pub struct Memorix {


}

const MEMORIX_NAMESPACE_NAME_TREE: &[&str] = &[];

impl Memorix {
    pub async fn new(redis_url: &str) -> Result<Memorix, Box<dyn std::error::Error>> {
        let memorix_base = memorix_client_redis::MemorixBase::new(
            redis_url,
            MEMORIX_NAMESPACE_NAME_TREE,
            Some(memorix_client_redis::MemorixOptions {
                cache: Some(memorix_client_redis::MemorixOptionsCache {
                    expire: Some(memorix_client_redis::MemorixOptionsCacheExpire {
                        value: 5,
                        is_in_ms: None,
                        extend_on_get: Some(true),
                    }),
                }),
                task: Some(memorix_client_redis::MemorixOptionsTask {
                    take_newest: Some(true),
                }),
            })
        ).await?;
        Ok(Self {


        })
    }
}"
`;

exports[`rust codegen enum can generate 1`] = `
"#![allow(dead_code)]
extern crate memorix_client_redis;

#[allow(non_camel_case_types, clippy::upper_case_acronyms)]
#[derive(Clone, memorix_client_redis::Serialize, memorix_client_redis::Deserialize, PartialEq, std::fmt::Debug)]
pub enum Animals {
    dog,
    cat,
    person,
}"
`;

exports[`rust codegen model can generate from 2 models 1`] = `
"#![allow(dead_code)]
extern crate memorix_client_redis;

#[derive(Clone, memorix_client_redis::Serialize, memorix_client_redis::Deserialize, PartialEq, std::fmt::Debug)]
pub struct User1 {
    pub id: i32,
}


#[derive(Clone, memorix_client_redis::Serialize, memorix_client_redis::Deserialize, PartialEq, std::fmt::Debug)]
pub struct User2 {
    pub name: String,
    #[serde(skip_serializing_if = \\"Option::is_none\\")]
    pub age: Option<i32>,
}"
`;

exports[`rust codegen model can generate from model 1`] = `
"#![allow(dead_code)]
extern crate memorix_client_redis;

#[derive(Clone, memorix_client_redis::Serialize, memorix_client_redis::Deserialize, PartialEq, std::fmt::Debug)]
pub struct User {
    pub id: i32,
    pub name: String,
    #[serde(skip_serializing_if = \\"Option::is_none\\")]
    pub age: Option<f32>,
}"
`;

exports[`rust codegen model can generate from model with array 1`] = `
"#![allow(dead_code)]
extern crate memorix_client_redis;

#[derive(Clone, memorix_client_redis::Serialize, memorix_client_redis::Deserialize, PartialEq, std::fmt::Debug)]
pub struct UserChildren {
    pub id: i32,
    #[serde(skip_serializing_if = \\"Option::is_none\\")]
    pub name: Option<String>,
}


#[derive(Clone, memorix_client_redis::Serialize, memorix_client_redis::Deserialize, PartialEq, std::fmt::Debug)]
pub struct User {
    pub id: i32,
    pub names: Vec<String>,
    #[serde(skip_serializing_if = \\"Option::is_none\\")]
    pub children: Option<Vec<Option<UserChildren>>>,
}"
`;

exports[`rust codegen model can generate from model within model 1`] = `
"#![allow(dead_code)]
extern crate memorix_client_redis;

#[derive(Clone, memorix_client_redis::Serialize, memorix_client_redis::Deserialize, PartialEq, std::fmt::Debug)]
pub struct UserPapa {
    #[serde(skip_serializing_if = \\"Option::is_none\\")]
    pub name: Option<String>,
}


#[derive(Clone, memorix_client_redis::Serialize, memorix_client_redis::Deserialize, PartialEq, std::fmt::Debug)]
pub struct User {
    pub id: i32,
    pub papa: UserPapa,
}"
`;

exports[`rust codegen model can generate from model within model within a model 1`] = `
"#![allow(dead_code)]
extern crate memorix_client_redis;

#[derive(Clone, memorix_client_redis::Serialize, memorix_client_redis::Deserialize, PartialEq, std::fmt::Debug)]
pub struct UserPapaMama {
    pub sick: bool,
}


#[derive(Clone, memorix_client_redis::Serialize, memorix_client_redis::Deserialize, PartialEq, std::fmt::Debug)]
pub struct UserPapa {
    #[serde(skip_serializing_if = \\"Option::is_none\\")]
    pub name: Option<String>,
    #[serde(skip_serializing_if = \\"Option::is_none\\")]
    pub mama: Option<UserPapaMama>,
    pub age: i32,
}


#[derive(Clone, memorix_client_redis::Serialize, memorix_client_redis::Deserialize, PartialEq, std::fmt::Debug)]
pub struct User {
    pub id: i32,
    pub papa: UserPapa,
}"
`;

exports[`rust codegen namespace can generate 1`] = `
"#![allow(dead_code)]
extern crate memorix_client_redis;

#[derive(Clone, memorix_client_redis::Serialize, memorix_client_redis::Deserialize, PartialEq, std::fmt::Debug)]
pub struct UserCacheBioPayload {
    pub description: String,
}


#[derive(Clone)]
#[allow(non_snake_case)]
pub struct MemorixCacheUser {
    pub bio: memorix_client_redis::MemorixCacheItemNoKey<UserCacheBioPayload>,
}

impl MemorixCacheUser {
    fn new(memorix_base: memorix_client_redis::MemorixBase) -> Self {
        Self {
            bio: memorix_client_redis::MemorixCacheItemNoKey::new(
                memorix_base.clone(),
                \\"bio\\".to_string(),
                None
            ),
        }
    }
}


#[derive(Clone)]
#[allow(non_snake_case)]
pub struct MemorixUser {

    pub cache: MemorixCacheUser,
}

const MEMORIX_USER_NAMESPACE_NAME_TREE: &[&str] = &[\\"user\\"];

impl MemorixUser {
    pub fn new(other: memorix_client_redis::MemorixBase) -> Result<MemorixUser, Box<dyn std::error::Error>> {
        let memorix_base = memorix_client_redis::MemorixBase::from(
            other,
            MEMORIX_USER_NAMESPACE_NAME_TREE,
            Some(memorix_client_redis::MemorixOptions {
                cache: Some(memorix_client_redis::MemorixOptionsCache {
                    expire: Some(memorix_client_redis::MemorixOptionsCacheExpire {
                        value: 5,
                        is_in_ms: None,
                        extend_on_get: None,
                    }),
                }),
                task: None,
            })
        );
        Ok(Self {

            cache: MemorixCacheUser::new(memorix_base.clone()),
        })
    }
}


#[derive(Clone)]
#[allow(non_snake_case)]
pub struct MemorixCacheSpaceship {
    pub bio: memorix_client_redis::MemorixCacheItemNoKey<String>,
}

impl MemorixCacheSpaceship {
    fn new(memorix_base: memorix_client_redis::MemorixBase) -> Self {
        Self {
            bio: memorix_client_redis::MemorixCacheItemNoKey::new(
                memorix_base.clone(),
                \\"bio\\".to_string(),
                None
            ),
        }
    }
}


#[derive(Clone)]
#[allow(non_snake_case)]
pub struct MemorixSpaceship {

    pub cache: MemorixCacheSpaceship,
}

const MEMORIX_SPACESHIP_NAMESPACE_NAME_TREE: &[&str] = &[\\"spaceship\\"];

impl MemorixSpaceship {
    pub fn new(other: memorix_client_redis::MemorixBase) -> Result<MemorixSpaceship, Box<dyn std::error::Error>> {
        let memorix_base = memorix_client_redis::MemorixBase::from(
            other,
            MEMORIX_SPACESHIP_NAMESPACE_NAME_TREE,
            Some(memorix_client_redis::MemorixOptions {
                cache: Some(memorix_client_redis::MemorixOptionsCache {
                    expire: Some(memorix_client_redis::MemorixOptionsCacheExpire {
                        value: 5,
                        is_in_ms: None,
                        extend_on_get: None,
                    }),
                }),
                task: None,
            })
        );
        Ok(Self {

            cache: MemorixCacheSpaceship::new(memorix_base.clone()),
        })
    }
}


#[derive(Clone)]
#[allow(non_snake_case)]
pub struct Memorix {
            pub user: MemorixUser,
            pub spaceship: MemorixSpaceship,


}

const MEMORIX_NAMESPACE_NAME_TREE: &[&str] = &[];

impl Memorix {
    pub async fn new(redis_url: &str) -> Result<Memorix, Box<dyn std::error::Error>> {
        let memorix_base = memorix_client_redis::MemorixBase::new(
            redis_url,
            MEMORIX_NAMESPACE_NAME_TREE,
            None
        ).await?;
        Ok(Self {
            user: MemorixUser::new(memorix_base.clone())?,
            spaceship: MemorixSpaceship::new(memorix_base.clone())?,


        })
    }
}"
`;

exports[`rust codegen pubsub can generate with inline types 1`] = `
"#![allow(dead_code)]
extern crate memorix_client_redis;

#[derive(Clone)]
#[allow(non_snake_case)]
pub struct MemorixPubSub {
    pub message: memorix_client_redis::MemorixPubSubItem<i32, String>,
}

impl MemorixPubSub {
    fn new(memorix_base: memorix_client_redis::MemorixBase) -> Self {
        Self {
            message: memorix_client_redis::MemorixPubSubItem::new(
                memorix_base.clone(),
                \\"message\\".to_string(),
            ),
        }
    }
}


#[derive(Clone)]
#[allow(non_snake_case)]
pub struct Memorix {

    pub pubsub: MemorixPubSub,
}

const MEMORIX_NAMESPACE_NAME_TREE: &[&str] = &[];

impl Memorix {
    pub async fn new(redis_url: &str) -> Result<Memorix, Box<dyn std::error::Error>> {
        let memorix_base = memorix_client_redis::MemorixBase::new(
            redis_url,
            MEMORIX_NAMESPACE_NAME_TREE,
            None
        ).await?;
        Ok(Self {

            pubsub: MemorixPubSub::new(memorix_base.clone()),
        })
    }
}"
`;

exports[`rust codegen task can generate with inline types 1`] = `
"#![allow(dead_code)]
extern crate memorix_client_redis;

#[derive(Clone)]
#[allow(non_snake_case)]
pub struct MemorixTask {
    pub doIt: memorix_client_redis::MemorixTaskItem<i32, String, bool>,
}

impl MemorixTask {
    fn new(memorix_base: memorix_client_redis::MemorixBase) -> Self {
        Self {
            doIt: memorix_client_redis::MemorixTaskItem::new(
                memorix_base.clone(),
                \\"doIt\\".to_string(),
                None
            ),
        }
    }
}


#[derive(Clone)]
#[allow(non_snake_case)]
pub struct Memorix {

    pub task: MemorixTask,
}

const MEMORIX_NAMESPACE_NAME_TREE: &[&str] = &[];

impl Memorix {
    pub async fn new(redis_url: &str) -> Result<Memorix, Box<dyn std::error::Error>> {
        let memorix_base = memorix_client_redis::MemorixBase::new(
            redis_url,
            MEMORIX_NAMESPACE_NAME_TREE,
            None
        ).await?;
        Ok(Self {

            task: MemorixTask::new(memorix_base.clone()),
        })
    }
}"
`;

exports[`rust codegen task can have options 1`] = `
"#![allow(dead_code)]
extern crate memorix_client_redis;

#[derive(Clone)]
#[allow(non_snake_case)]
pub struct MemorixTask {
    pub doIt: memorix_client_redis::MemorixTaskItem<i32, String, bool>,
}

impl MemorixTask {
    fn new(memorix_base: memorix_client_redis::MemorixBase) -> Self {
        Self {
            doIt: memorix_client_redis::MemorixTaskItem::new(
                memorix_base.clone(),
                \\"doIt\\".to_string(),
                Some(memorix_client_redis::MemorixOptionsTask {
                    take_newest: Some(true),
                })
            ),
        }
    }
}


#[derive(Clone)]
#[allow(non_snake_case)]
pub struct Memorix {

    pub task: MemorixTask,
}

const MEMORIX_NAMESPACE_NAME_TREE: &[&str] = &[];

impl Memorix {
    pub async fn new(redis_url: &str) -> Result<Memorix, Box<dyn std::error::Error>> {
        let memorix_base = memorix_client_redis::MemorixBase::new(
            redis_url,
            MEMORIX_NAMESPACE_NAME_TREE,
            None
        ).await?;
        Ok(Self {

            task: MemorixTask::new(memorix_base.clone()),
        })
    }
}"
`;
